### DISCUSSION TITLE
Backup creates empty files

### DISCUSSION THREAD

**Discussion by KnightwhosayNi777** [teslamate-org/teslamate#2138] [Q&A] [answered]
Hello

I use the following script to backup my Teslamate database. But unfortunately an error has crept in, the backup file is empty, 0 bytes.

Does anyone have an idea why? 

```
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
now=`date +"%A"`
cd /home/pi/tmbackup
sudo /usr/local/bin/docker-compose exec -T database pg_dump -U teslamate teslamate > /home/pi/tmbackup/teslamate.bck_${now}

```
The script runs without errors, the file teslamate.bck_Monday is created as expected. However, when I check the size of teslamate.bck_Monday, I see the file is empty.

```
root@raspi4:/home/pi/tmbackup# ls -l
total 4
-rw-r--r-- 1 root root   0 Nov  8 12:59 teslamate.bck_Monday
-rwxr-xr-x 1 root root 429 Nov  8 12:59 tmbackup.sh
```
Is docker-compose command wrong or why does the backup not contain the teslamate database?

**Comment by tucopup** [comment]
I had this issue as well.  It was some time ago, so I'm not sure.

I recall that chmod 777 on the target folder fixed it.  Or the script running as root.  One or the other or both.

Teslamate runs on a Synology NAS in Docker.  Here is the script (database name may differ):

cd /volume1/@docker
docker-compose exec -T db pg_dump -U teslamate teslamate > /volume1/Tom/Teslamate/teslamate.bck_`date +%d%b%Y%H%M%S`


Another script limits the backups in the folder to 30 (also running as root)

find /volume1/Tom/Teslamate -type f -mtime +30 -exec rm -f {} \;

**Reply by KnightwhosayNi777** [reply]
I have tried both, but it still does not work.

The files are still stored on qnap (rclone), which does not work yet, but does not matter.

```
pi@raspi4:~ $ sudo chmod 777 /home/pi//tmbackup/
pi@raspi4:~ $ cd /home/pi/tmbackup/
pi@raspi4:~/tmbackup $ ./tmbackup.sh 
2021/11/09 19:45:01 NOTICE: Config file "/home/pi/.config/rclone/rclone.conf" not found - using defaults
2021/11/09 19:45:01 Failed to create file system for "qnap:Backups/BackupTeslaMate": didn't find section in config file
pi@raspi4:~/tmbackup $ ls -l
total 4
-rw-r--r-- 1 root root   0 Nov  8 13:31 teslamate.bck_Monday
-rw-r--r-- 1 pi   pi     0 Nov  9 19:44 teslamate.bck_Tuesday
-rwxr-xr-x 1 root root 325 Nov  8 13:06 tmbackup.sh
pi@raspi4:~/tmbackup $ sudo ./tmbackup.sh 
pi@raspi4:~/tmbackup $ ls -sl
total 4
0 -rw-r--r-- 1 root root   0 Nov  8 13:31 teslamate.bck_Monday
0 -rw-r--r-- 1 pi   pi     0 Nov  9 19:45 teslamate.bck_Tuesday
4 -rwxr-xr-x 1 root root 325 Nov  8 13:06 tmbackup.sh
```

**Comment by refrain** [comment]
Run just the command in a shell and do you get any error messages?

```
cd /home/pi/tmbackup   # assuming your teslamate folder is /home/pi/tmbackup that contains the docker-compose.yml ?
sudo docker-compose exec -T database pg_dump -U teslamate teslamate > /home/pi/tmbackup/teslamate.bck
```

Is docker-compose supposed to be in `/usr/local/bin`?  Did you try typing

`whereis docker-compose`

Normally it should be in `/usr/bin` unless you compile it yourself or something?

If it's a permission issue, you need to use:

`sudo sh -c "docker-compose exec -T database pg_dump -U teslamate teslamate > /home/pi/tmbackup/teslamate.bck"`

Also ideally you are putting your backups in your SSD not your `/home` folder which is on your sd card...

**Reply by KnightwhosayNi777** [reply]
That's what I executed and what came out of it. 


```
pi@raspi4:~/tmbackup $ ./tmbackup.sh   #Command in a shell
pi@raspi4:~/tmbackup $ ls -l
total 4
-rw-r--r-- 1 pi   pi     0 Nov  9 19:56 teslamate.bck
-rw-r--r-- 1 root root   0 Nov  8 13:31 teslamate.bck_Monday
-rw-r--r-- 1 pi   pi     0 Nov  9 20:04 teslamate.bck_Tuesday
-rwxr-xr-x 1 root root 326 Nov  9 20:04 tmbackup.sh
```

```
pi@raspi4:~/tmbackup $ whereis docker-compose
docker-compose: /usr/local/bin/docker-compose
```
```
pi@raspi4:~/tmbackup $ sudo sh -c "docker-compose exec -T database pg_dump -U teslamate teslamate > /home/pi/tmbackup/teslamate.bck"
pi@raspi4:~/tmbackup $ ls -l
total 4
-rw-r--r-- 1 pi   pi     0 Nov  9 19:56 teslamate.bck
-rw-r--r-- 1 root root   0 Nov  8 13:31 teslamate.bck_Monday
-rw-r--r-- 1 pi   pi     0 Nov  9 19:45 teslamate.bck_Tuesday
-rwxr-xr-x 1 root root 325 Nov  8 13:06 tmbackup.sh
```

The file he created again but is still empty. 

My Raspberry has a SD card and an external USB HDD. Docker will run on the HDD, the local backup files on the SD card and a copy on Qnap. So all data is on three different storages.

**Reply by refrain** [reply]
The problem is simple...  docker-compose needs to be run in the same directory as your docker-compose.yml file.  Your script needs to cd to your teslamate folder that contains the yml file then run the command.

**Comment by KnightwhosayNi777** [comment, marked as answer]
Thx @refrain, you were right. Script needs to cd to teslamate folder that contains the yml file.

```
pi@raspi4:~/tmbackup` $ ./tmbackup.sh 
pi@raspi4:~/tmbackup $ ls -l
total 322996
-rw-r--r-- 1 pi   pi           0 Nov  9 19:56 teslamate.bck
-rw-r--r-- 1 root root         0 Nov  8 13:31 teslamate.bck_Monday
-rw-r--r-- 1 pi   pi   330737180 Nov  9 20:18 teslamate.bck_Tuesday
-rwxr-xr-x 1 root root       329 Nov  9 20:17 tmbackup.sh
```

**Comment by static-glitch** [comment]
I'm getting empty files as well. I'm running the following command `docker-compose exec -T database pg_dump -U teslamate teslamate > teslamate.bck` in the directory where my compose.yml is. I've tried subbing in the name of the database image which in my case is "docker-files_database_1" (I'm a docker noob and every few months that I need to interact with my containers, I learn something, often the hard way). I've moved my compose folder since I initially created the containers and that made me think I lost my data when upgrading/recreating, so I'm wanting to backup my data one, because I should, but two so I can recreate my images with better names that aren't dependent on the compose file's parent directory.

I get no errors, just an empty file.

Here's my current filestructure: 
```
media@mediaserver:~/docker-files/teslamate$ docker-compose exec -T docker-files_database_1 pg_dump -U docker-files_teslamate_1 teslamate > teslamate.bck
media@mediaserver:~/docker-files/teslamate$ ll
total 16
drwxrwxr-x  3 media media 4096 Jul 25 22:56 ./
drwxrwxr-x 10 media media 4096 May  5 00:08 ../
lrwxrwxrwx  1 media media   21 Jul 25 22:56 compose.yml -> teslamate-compose.yml
drwxr-xr-x  2 root  root  4096 Jul 24 21:51 import/
-rw-rw-r--  1 media media    0 Jul 25 23:10 teslamate.bck
-rw-rw-r--  1 media media 1529 Jul 25 22:49 teslamate-compose.yml
```
How can I get a proper backup completed?

**Reply by cwanja** [reply]
Total Docker noob myself. But I think it needs to be named docker-compose.yml? Also not sure symlinks would work, if you were planning to move the file eventually.

**Reply by static-glitch** [reply]
From what I've read, docker-compose.yml, docker-compose.yaml, compose.yml, and compose.yaml are all interchangable and what docker-compose looks for by default. The only reason I've got the symlink is so I can run the backup command with the compose named differently (and without specifying the compose file with -f because I'm not sure it works in this context). When I originally started the container, it and several others compose files were all in `~/docker-files/` and then I moved them to their own folders in that directory such as `~/docker-files/teslamate/`

I'm thinking this move might be why it's not working at the moment. I'm at work and can't move the compose and retest until later. Unless someone else knows better?

**Reply by static-glitch** [reply]
Aha! I think it was because my currently running docker image for TeslaMate was created from my `~/docker-files/` directory and thus named `docker-files_teslamate_1` but specifying the container name in the `docker-compose exec` command didn't do the trick. To be honest, I don't know what each positional parameter is supposed to be in that command. Anyway, `media@mediaserver:~/docker-files$ cp teslamate/* .` (~~honestly not sure if I needed to do this or not~~ I tested, it was needed) followed by running the backup command from `~/docker-files/` yielded a 166M file. Now I can spin it up with a proper name in it's own directory!
